sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedPartitions.scala

```diff
@@ -0,0 +1,259 @@
+              getNumMappers(right)
+            } else {
+              rightMapIdStartIndices(j + 1)
+            }
+            // TODO: we may can optimize the sort merge join to broad cast join after
+            //       obtaining the raw data size of per partition,
+            val leftSkewedReader = SkewedShufflePartitionReader(
+              left, partitionId, leftMapIdStartIndices(i), leftEndMapId)
+
+            val rightSkewedReader = SkewedShufflePartitionReader(right, partitionId,
+                rightMapIdStartIndices(j), rightEndMapId)
+            subJoins += SortMergeJoinExec(leftKeys, rightKeys, joinType, condition,
+              leftSkewedReader, rightSkewedReader)
+          }
+        }
+      }
+      logDebug(s"number of skewed partitions is ${skewedPartitions.size}")
+      if (skewedPartitions.size > 0) {
+        val optimizedSmj = smj.transformDown {
+          case sort: SortExec if (sort.child.isInstanceOf[ShuffleQueryStageExec]) =>
+            val shuffleStage = sort.child.asInstanceOf[ShuffleQueryStageExec]
+            val newStage = shuffleStage.copy(
+              excludedPartitions = skewedPartitions.toSet)
+            newStage.resultOption = shuffleStage.resultOption
+            sort.copy(child = newStage)
+        }
+        subJoins += optimizedSmj
+        UnionExec(subJoins)
+      } else {
+        smj
+      }
+  }
+
+  override def apply(plan: SparkPlan): SparkPlan = {
+    if (!conf.getConf(SQLConf.ADAPTIVE_EXECUTION_SKEWED_JOIN_ENABLED)) {
+      return plan
+    }
+
+    def collectShuffleStages(plan: SparkPlan): Seq[ShuffleQueryStageExec] = plan match {
+      case _: LocalShuffleReaderExec => Nil
+      case _: CoalescedShuffleReaderExec => Nil
+      case stage: ShuffleQueryStageExec => Seq(stage)
+      case _ => plan.children.flatMap(collectShuffleStages)
+    }
+
+    val shuffleStages = collectShuffleStages(plan)
+
+    if (shuffleStages.distinct.length == 2) {
+      // Currently we only support handling skewed join for 2 table join.
+      handleSkewJoin(plan)
+    } else {
+      plan
+
+    }
+  }
+}
+
+/**
+ * A wrapper of shuffle query stage, which submits one reduce task to read a single
+ * shuffle partition 'partitionIndex' produced by the mappers in range [startMapId, endMapId).
+ * This is used to handle the skewed partitions.
+ *
+ * @param child It's usually `ShuffleQueryStageExec` or `ReusedQueryStageExec`, but can be the
+ *              shuffle exchange node during canonicalization.
+ * @param partitionIndex The pre shuffle partition index.
+ * @param startMapId The start map id.
+ * @param endMapId The end map id.
+ */
+case class SkewedShufflePartitionReader(
+    child: QueryStageExec,
+    partitionIndex: Int,
+    startMapId: Int,
+    endMapId: Int) extends LeafExecNode {
+
+  override def output: Seq[Attribute] = child.output
+
+  override def doCanonicalize(): SparkPlan = child.canonicalized
```

Comment:
We shouldn't override it. `SkewedShufflePartitionReader` produces different result than `child`